pkg/kubelet/prober/worker.go

```diff
@@ -218,6 +221,11 @@ func (w *worker) doProbe() (keepGoing bool) {
 		return true
 	}
 
+	// Record the end of the initialization phase.
+	if result == results.Success && !w.hasInitialized {
```

Comment:
We need to get someone from SIG Node to weigh in on this.

* Right now, this code might randomly allow MaxInitialFailureCount failures after the pod has become alive, if kubelet happens to restart at the same time the pod goes unhealthy. Is that OK from a UX perspective? From a Kubelet perspective? (i.e., do other things already behave like that?)

Reply:
Normally, kubelet handles container/pod lifecycle-related functions by relying on the underlying container runtime to persist the states. For probes and/or lifecycle hooks, kubelet rely on in-memory states only.

This is consistent with the existing probe behavior, but the existing behavior does not care about pod lifecycle much. If the MaxInitialFailureCount is set very high, it can potentially make the container downtime much longer and will need to be used with caution (documentation?).

By the way, just to make it clear, even though we use the term "pod" in the discussion threads, the changes pertain to "containers" and "container starts".

Reply:
@yujuhong maybe we could persist the `hasInitialized` state somewhere inside the container itself? Why not a file that we create somewhere?
But it would do an exec every time we need to check that state...

Reply:
> If the MaxInitialFailureCount is set very high, it can potentially make the container downtime much longer and will need to be used with caution (documentation?).

Indeed, but people in need of that feature would normally use a very high FailureThreshold, so it wouldn't be worse in that case...

Reply:
Detailed answer here: https://github.com/kubernetes/enhancements/blob/4023cd9b30b33abe9b36b8f3826c13d02e92a450/keps/sig-node/20190221-maxInitialFailureCount-health-probes.md#stateless-kubelet